### ISSUE TITLE
LeechDataset holds three copies of the chunk data at load: a 41 GB npz peaks at 116 GB and OOMs before epoch 1

### ISSUE THREAD

**Issue by jayhesselberth** [rnabioco/leech#211]
## What happens

`LeechDataset.__init__` materialises the same data three times, and all three are alive at the peak:

1. **`load_chunks()`** (`dataset.py:277`) reads every member of the npz with `np.load` — 41.1 GB for the file below. The per-chunk dicts hold **views** into those arrays, so none of it can be freed while `self.chunks` is alive.
2. **The tensorize loop** (`dataset.py:333`) builds one tensor per chunk into `self._signals` / `self._features`. `_prepare_signal` crops/pads and stacks signal+residual, so these are copies, not views: another ~36 GB.
3. **`_try_stack`** (`dataset.py:399`) calls `torch.stack(items)`, which allocates the whole contiguous output *while the list is still alive*: another ~36 GB, transient.

The raw arrays are nulled out of `self.chunks` at `dataset.py:465`, for exactly the right reason (the COW blowup on DataLoader fork) — but that runs **after** step 3, so it does nothing for the peak.

## Measured

leech 0.6.7, `TCNDwellResidualLN`, `signal_mode: both`, `--num-workers 2` (so this is not the fork blowup). Four arms of one sweep differing only in `--signal-context 90 R`. `sacct MaxRSS`:

| right context | train.npz + val.npz | MaxRSS | outcome |
|---|---|---|---|
| 150 | 22.9 GB | 72.8 / 70.6 GB | ok |
| 300 | 38.7 GB | 99.7 / 104.4 GB | ok |
| 450 | 48.3 GB | **116 GB** | OUT_OF_MEMORY at 8m27s, 120 GB cap |

Both 450 arms died at 8m27s — during the load, before the first epoch. RSS runs ~2.4–2.6x the file size, and the slope against right context is ~0.2 GB per sample.

The npz in question (6,668,328 chunks), read from the zip headers without loading it:

```
signals_flat                (6668328, 540)     float32  14.40 GB  stored
signal_residuals_flat       (6668328, 540)     float32  14.40 GB  stored
features_flat               (6668328, 12, 21)  float32   6.72 GB  stored
sequences_with_kmer_context (6668328,)         <U56      1.49 GB  stored
seq_to_sig_maps             (6668328,)         object    0.90 GB  stored
... 12 more 1-D members                                  3.17 GB
TOTAL                                                   41.08 GB
```

## Why it costs more than the one OOM

Our GPU nodes have 251 GB and four A30s. The training rule requests 120 GB *precisely* so two arms share a node and two GPUs stay busy; at 160 GB only one fits. So raising the request to fit the widest arm serialises the sweep and idles three of four GPUs — the copies cost throughput on every arm that does fit, not just on the one that doesn't.

## Suggested fixes, cheapest first

1. **Preallocate and fill instead of `torch.stack`.** The output shape is known before the loop (`n_chunks x channels x signal_len`), so `out = torch.empty(...)` plus `out[i] = ...` removes the ~36 GB transient outright. Same for features. Local, no format change.

2. **Convert in row blocks and never hold the numpy source.** `signals_flat`, `signal_residuals_flat`, `features_flat` and `dwells_flat` are fixed-shape members, and everything `_prepare_signal` does to them is a crop/pad that is expressible on a block of rows. Reading them block-wise (sequential IO, which is the friendly pattern on a network filesystem) straight into the preallocated tensor takes the peak to about the size of the output alone — roughly 36 GB instead of 116 GB here. `focus_signal_pos` varies per chunk so the crop is a gather, but a gather per block is still vectorised.

3. **`load_chunks`'s docstring is wrong about this** (`chunking/serialization.py:173-193`): "The loaded data is kept in memory-mapped form when possible". `np.load` never mmaps a zip member, so it is always a full read. Worth correcting, because it invites the assumption that this path is already lazy — it is what we assumed before measuring.

4. **Longer term, if a genuinely mappable file is ever wanted:** under `--no-compress` every member is written *stored*, so all the fixed-shape ones could be `np.memmap`'d in place at their zip offsets and `__getitem__` could slice from disk. The blocker is `seq_to_sig_maps`, a pickled object array; storing it as flat values + offsets would make the whole file mappable. Caveat from our side: shuffled random access into a mapped file is only viable while the page cache holds it (our BeeGFS does ~113 random IOPS per thread against a loader that wants ~10,900 chunks/s), so this is a design question rather than a drop-in — unlike 1 and 2.

Fixes 1 and 2 need no on-disk format change, so existing corpora stay valid.

**Comment by jayhesselberth** [comment]
A second instance of the same root cause, found while re-running the arms that OOM'd: **the fork-safety stacking is conditional, so models outside `FEATURE_MODELS` keep the exact object graph the stacking exists to avoid.**

`dataset.py:447-451`:

```python
self._features_tensor: torch.Tensor | None = None
if self._needs_features:
    self._features_tensor = _try_stack("Feature", self._features)
    if self._features_tensor is not None:
        self._features = []
```

For a model that doesn't need features (`ConvLSTMBase` here), `_needs_features` is False, so `self._features` stays a list of **N `torch.empty(0)` tensors** — 7,599,113 of them on this test set — never stacked, never cleared, and still indexed by `__getitem__` (`dataset.py:852`). Eight forked DataLoader workers then touch every one of those refcounts, which is precisely the COW blowup your comment above `_try_stack` describes.

Measured on one `test.npz`, same geometry, same data, `leech eval test`:

| model | MaxRSS | outcome |
|---|---|---|
| TCNDwellResidualLN | 114.5 / 114.8 GB | passed |
| ConvLSTMBase | >122.7 GB | OOM at a 120 GB cap |
| ConvLSTMBase | **143.5 GB** | OOM again at a 150 GB cap |

So the *lighter* model — signal + sequence, no feature branch — costs ~30 GB more than the heavier one, purely because its unused branch is the one left un-stacked. Dropping `--num-workers` from the auto-8 to 2 is what let it run here.

Suggestion, alongside the three-copies fixes: stack (or simply drop) **every** per-chunk list unconditionally, including the branches the model won't read. `_features` for a non-feature model can be a single `torch.empty(N, 0)` or a `None` that `__getitem__` special-cases — either way there should be no path that leaves N live Python objects behind a `DataLoader` fork. The same guard shape appears for `_confound_labels` (`if self._has_confound`) and `_cl_targets` (`if self._cl_regression`), so those look like the same hazard when the flags are off.
